pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go

```diff
@@ -319,36 +278,185 @@ func (grm *nestedPendingOperations) operationComplete(key operationKey, err *err
+			}
+		}
+		nro := vro.nodeRelated[opKey.nodeName]
+		if opKey.podName != EmptyUniquePodName {
+			pro := &podRelatedOperation{
+				operationName: operName,
+			}
+			nro.podRelated[opKey.podName] = pro
+			return
+		}
+		nro.operationName = operName
+		return
+	}
+	// podName and nodeName both empty
+	vro.operationName = operName
+}
+
+func (o *operationsCache) addOperation(op *operation) {
+	// Assumes lock has been acquired by caller.
+	opKey := op.key
+	// Add operation
+	o.allOps[opKey] = op
+	// Add pending recored
+	o.addPending(opKey)
+}
+
+func (o *operationsCache) getOperation(opKey operationKey) (*operation, error) {
+	// Assumes lock has been acquired by caller.
+	if op, ok := o.allOps[opKey]; ok {
+		return op, nil
+	}
+	return nil, fmt.Errorf("operation %+v not found", opKey)
+}
+
+func (o *operationsCache) deletePending(opKey operationKey) {
+	// Remove pending records
+	if volumePendingOp, ok := o.pendingOps[opKey.volumeName]; ok {
+		if opKey.podName != EmptyUniquePodName {
+			if nodePendingOp, ok := volumePendingOp.nodeRelated[opKey.nodeName]; ok {
+				delete(nodePendingOp.podRelated, opKey.podName)
+				if len(nodePendingOp.podRelated) == 0 {
+					delete(volumePendingOp.nodeRelated, opKey.nodeName)
+				}
+			}
+		} else if opKey.nodeName != EmptyNodeName {
+			delete(volumePendingOp.nodeRelated, opKey.nodeName)
+		} else {
+			volumePendingOp.operationName = ""
+		}
+		// Cleanup pendings
+		if len(volumePendingOp.operationName) == 0 && len(volumePendingOp.nodeRelated) == 0 {
+			delete(o.pendingOps, opKey.volumeName)
+		}
+	}
+}
+
+func (o *operationsCache) deleteOperation(opKey operationKey) {
+	// Assumes lock has been acquired by caller.
+
+	// Remove pending records
+	o.deletePending(opKey)
+	// Remove operation
+	delete(o.allOps, opKey)
+}
+
+func (o *operationsCache) updateOperation(opKey operationKey, isPending, isUpdateBackoff bool, backOffErr *error) {
+	// Assumes lock has been acquired by caller.
+	if op, ok := o.allOps[opKey]; ok {
+		if isUpdateBackoff {
+			op.expBackoff.Update(backOffErr)
+		}
+		op.operationPending = isPending
+	}
+	if isPending {
+		o.addPending(opKey)
+	} else {
+		o.deletePending(opKey)
+	}
+}
+
+func (o *operationsCache) isOperationExists(opKey operationKey) (bool, *operation) {
+	// Assumes lock has been acquired by caller.
+	if op, ok := o.allOps[opKey]; ok {
+		return ok, op
+	}
+	return false, nil
+}
+
+func (o *operationsCache) hasOperationPending(volumeName v1.UniqueVolumeName,
+	podName volumetypes.UniquePodName,
+	nodeName types.NodeName) bool {
+	// Assumes lock has been acquired by caller.
+
+	vro, exists := o.pendingOps[volumeName]
+	if !exists {
+		return false
+	}
+	if volumeName == EmptyUniqueVolumeName {
+		// volumeName empty, nodeName should exists(for verify_volumes_are_attached_per_node)
+		// to make the opKey unique. if the operation for the same node is still pending,
+		// should reject current operation?
```

Comment:
right now only verifyVolumeAttached function uses empty volumeName to allow it to run parallel no matter what other volume operations are running. I think we can still make this if part at the beginning of the function